### PR TITLE
Upgrade to LLVM 13

### PR DESCRIPTION
**Pull request by nikic**
Work in progress update to LLVM 13. Main changes:

 * InlineAsm diagnostics reported using SrcMgr diagnostic kind are now handled. Previously these used a separate diag handler.
 * Codegen tests are updated for additional attributes.
 * Some data layouts have changed.
 * Switch `#[used]` attribute from `llvm.used` to `llvm.compiler.used` to avoid SHF_GNU_RETAIN flag introduced in https://reviews.llvm.org/D97448, which appears to trigger a bug in older versions of gold.
 * Set `LLVM_INCLUDE_TESTS=OFF` to avoid Python 3.6 requirement.

Upstream issues:

 * ~~https://bugs.llvm.org/show_bug.cgi?id=51210 (InlineAsm diagnostic reporting for module asm)~~ Fixed by https://github.com/llvm/llvm-project/commit/1558bb80c01b695ce12642527cbfccf16cf54ece.
 * ~~https://bugs.llvm.org/show_bug.cgi?id=51476 (Miscompile on AArch64 due to incorrect comparison elimination)~~ Fixed by https://github.com/llvm/llvm-project/commit/81b106584f2baf33e09be2362c35c1bf2f6bfe94.
 * https://bugs.llvm.org/show_bug.cgi?id=51207 (Can't set custom section flags anymore). Problematic change reverted in our fork, https://reviews.llvm.org/D107216 posted for upstream revert.
 * https://bugs.llvm.org/show_bug.cgi?id=51211 (Regression in codegen for #83623). This is an optimization regression that we may likely have to eat for this release. The fix for #83623 was based on an incorrect premise, and this needs to be properly addressed in the MergeICmps pass.

The [compile-time impact](https://perf.rust-lang.org/compare.html?start=ef9549b6c0efb7525c9b012148689c8d070f9bc0&end=0983094463497eec22d550dad25576a894687002) is mixed, but quite positive as LLVM upgrades go.

The LLVM 13 final release is scheduled for Sep 21st. The current nightly is scheduled for stable release on Oct 21st.

r? @ghost

### CODE REVIEW

**Comment by nikic** [comment]
It looks like the large regressions at the start are all slowdowns in run_linker, so this might be related to the missing llvmbc section flag.

**Comment by hkratz** [comment]
> It looks like the large regressions at the start are all slowdowns in run_linker, so this might be related to the missing llvmbc section flag.

Seems like it. The resulting binaries do get bloated with bitcode. 

The resulting stripped binary of compiling the `helloworld` perf test...

with Rust nightly 2021-07-28:
```
rustc -C opt-level=3 src/main.rs;strip main;ls -l main
-rwxr-xr-x 1 hans hans 272656 Jul 29 09:54 main
```

with LLVM 13:
```
~/dev/rust/build/x86_64-unknown-linux-gnu/stage2/bin/rustc -C opt-level=3 src/main.rs;strip main;ls -l main
-rwxr-xr-x 1 hans hans 6384232 Jul 29 10:07 main
```

Just backing out the hack and reverting the LLVM commit to allow the SHF_EXCLUDE flag on .llvmbc again works and makes it better:
```
~/dev/rust/build/x86_64-unknown-linux-gnu/stage2/bin/rustc -C opt-level=3 src/main.rs;strip main;ls -l main
-rwxr-xr-x 1 hans hans 309520 Jul 29 09:56 main
```

**Comment by nikic** [comment]
Weird issue encountered via the `pgo-branch-weights` test:
```
fn main() {
    println!("{:?}", std::env::args());
}
```
Running `rustc test.rs && ./test` works fine. Running `rustc -Clink-args=-fuse-ld=gold test.rs && ./test` gives empty arguments.

**Comment by nikic** [comment]
> Just backing out the hack and reverting the LLVM commit to allow the SHF_EXCLUDE flag on .llvmbc again works and makes it better:
> 
> ```
> ~/dev/rust/build/x86_64-unknown-linux-gnu/stage2/bin/rustc -C opt-level=3 src/main.rs;strip main;ls -l main
> -rwxr-xr-x 1 hans hans 309520 Jul 29 09:56 main
> ```

Unfortunately, it seems like this doesn't actually work: Due to https://reviews.llvm.org/D100944 this no longer modifies the section flags on `.llvmbc` and will instead emit first an empty `.llvmbc` section with EXCLUDE, and then the actual `.llvmbc` section with ALLOC. LLVM is then no longer able to find the bitcode.

**Comment by nikic** [comment]
The good news is that reverting (just) https://reviews.llvm.org/D100944 does fix the llvmbc issue.

> Running `rustc test.rs && ./test` works fine. Running `rustc -Clink-args=-fuse-ld=gold test.rs && ./test` gives empty arguments.

It looks like something is going wrong with `.init_array`:

```
objdump -t test_ld | grep init_array
0000000000048250 l    d  .init_array	0000000000000000              .init_array
0000000000048258 l     O .init_array	0000000000000000              __frame_dummy_init_array_entry
0000000000048250 l     O .init_array	0000000000000008              _ZN3std3sys4unix4args3imp15ARGV_INIT_ARRAY17hb13aa0f306b867eeE
0000000000048260 l       .init_array	0000000000000000              __init_array_end
0000000000048250 l       .init_array	0000000000000000              __init_array_start

objdump -t test_gold | grep init_array
000000000004a218 l     O .init_array	0000000000000000              __frame_dummy_init_array_entry
000000000004c728 l     O .init_array	0000000000000008              _ZN3std3sys4unix4args3imp15ARGV_INIT_ARRAY17hb13aa0f306b867eeE
000000000004a218 l       .init_array	0000000000000000              .hidden __init_array_start
000000000004a220 l       .init_array	0000000000000000              .hidden __init_array_end
```

Notably, with `ld` we have `ARGV_INIT_ARRAY` sitting between `__init_array_start` and `__init_array_end`. With `gold` it's located after `__init_array_end`.

**Comment by nikic** [comment]
Still rather stumped on the gold issue and don't really know how to debug it.

The best lead I have right now is that there are now some adjacent anon globals in the `.data.rel.ro` section, which were not present previously. Extracting the bitcode from the stdlib rlib, the relevant difference seems to be that a number of panic message constants are now `hidden unnamed_addr constant`, while they were previously `private unnamed_addr constant`. This means that these symbols now have external linkage, which they shouldn't have.

**Comment by tmiasko** [comment]
> Running `rustc test.rs && ./test` works fine. Running `rustc -Clink-args=-fuse-ld=gold test.rs && ./test` gives empty arguments.

`#[used]` attribute now results in `SHF_GNU_RETAIN` / `R` flag being set https://reviews.llvm.org/D97448, which seems to be causing the issue when linking with gold < 2.36.

**Comment by nikic** [comment]
@tmiasko Ooh, nice find!

As suggested in the differential, we might want to switch to `llvm.compiler.used` then, as all our docs on `#[used]` make it abundantly clear that the attribute is only supposed to ensure that the symbol makes it into the object file, but the linker is still allowed to drop it.

**Comment by nikic** [comment]
Switching from llvm.used to llvm.compiler.used makes tests pass locally.

**Comment by nikic** [comment]
I've tested a handful of docker images locally without encountering further issues, so I think this is ready for review.

r? @nagisa or @cuviper

**Comment by nagisa** [comment]
In response to the `used` semantics, we ended up specifying `#[used]` as requiring the compiler to retain it, but not necessarily the linker, but that has long time been a problematic take with people struggling to understand why this distinction is necessary. I imagine we might want to consider adopting the new implementation for `#[used]` for at least GNU platforms (without changing the definition).

File: compiler/rustc_codegen_llvm/src/context.rs [hunk 1]
```diff
@@ -118,6 +118,14 @@ pub unsafe fn create_module(
...
         target_data_layout = strip_powerpc64_vectors(target_data_layout);
     }
+    if llvm_util::get_version() < (13, 0, 0) {
+        if sess.target.arch == "wasm32" {
+            target_data_layout = "e-m:e-p:32:32-i64:64-n32:64-S128".to_string();
+        }
+        if sess.target.arch == "wasm64" {
+            target_data_layout = "e-m:e-p:64:64-i64:64-n32:64-S128".to_string();
+        }
+    }
```
Comment (nagisa) [inline compiler/rustc_codegen_llvm/src/context.rs:127]: This seems pretty poor, if we're resorting to doing this we might as well not include the data layout to the target definition at all. 

Out of scope for this PR regardless, though.

Comment (nikic) [reply compiler/rustc_codegen_llvm/src/context.rs:127]: I believe the reason the target data layout exist is that rustc needs the information itself for layout calculation and some other things.

But we don't need all the information -- only the parts parsed into TargetDataLayout. So possibly what we could do is reduce the data layout info in targets to just the necessary information, and then only check that this is consistent with the LLVM data layout (but then use the more precise LLVM data layout for code generation).

In these two cases, the newly added information is stack alignment and non-integral address spaces, both of which we don't care about.

**Comment by nikic** [comment]
Applied a fix for the mingw build failure in https://github.com/rust-lang/llvm-project/commit/5f278975dee31dd0bfba0241b19210c4fa058e4b, will upstream if it works.

@bors r=nagisa

**Comment by nikic** [comment]
Applied fix for FreeBSD build in https://github.com/rust-lang/llvm-project/commit/129a736f429dfc320080f15e20a202b4271a55c0. This should also be upstreamed, as it is a supported compiler version.

@bors r=nagisa

**Comment by nikic** [comment]
Cherry-picked https://github.com/rust-lang/llvm-project/commit/94124e25631aa95a9dde1f0f118944634811951d to fix the armhf failure. This one is already upstream, but not yet backported.

@bors r=nagisa

**Comment by nikic** [comment]
`rustdoc-ui/intra-doc/errors.rs` failed on aarch64-gnu, presumably an AArch64 miscompile.

**Comment by nikic** [comment]
The problem seems to be that this loop gets skipped: https://github.com/rust-lang/rust/blob/5de331ba40ac722dd5513c79194c6c411766a23f/src/librustdoc/passes/collect_intra_doc_links.rs#L1382

```
  250684:       7100051f        cmp     w8, #0x1
  250688:       1a9f07e1        cset    w1, ne  // ne = any
  25068c:       32170028        orr     w8, w1, #0x200
  250690:       3d8002e0        str     q0, [x23]
  250694:       790a23e8        strh    w8, [sp, #1296]
  250698:       5400a4c1        b.ne    251b30 <_ZN7rustdoc6passes23collect_intra_doc_links13LinkCollector12resolve_link17h9b60d95544143a64E+0x42cc>  // b.any
```

w1 is other_ns here, w8 are the array contents (other_ns + MacroNS) and then the `b.ne` skips over the whole loop if expected_ns is TypeNS. Not sure where that branch is coming from though.

**Comment by nikic** [comment]
https://bugs.llvm.org/show_bug.cgi?id=51476 for the AArch64 miscompile.

**Comment by nikic** [comment]
Cherry-picked https://github.com/rust-lang/llvm-project/commit/609f806042a0f290a2f9a386d5ba420f6ae6fc2b for AArch64 miscompile, and also merged the current release/13.x branch.

@bors r=nagisa

**Comment by nikic** [comment]
I wasn't able to reproduce the segfault, but I did see a use-after free under valgrind, which is now fixed. Let's see if that also fixes the segfault in CI...

Somewhat surprised we never ran into this issue before.

@bors r=nagisa

**Comment by nikic** [comment]
mingw `__llvm_profile_counter_bias` link failure is probably caused by https://reviews.llvm.org/D105176.

**Comment by nikic** [comment]
Following https://reviews.llvm.org/D107620 pass `-fms-extensions` when building libprofile for mingw. I don't think this is actually going to work, because I believe this requires clang while we build with gcc. But let's give it a...

@bors try

**Comment by nikic** [comment]
As expected, that didn't work. Let's see whether the non-windows code path happens to work with mingw.

@bors try

**Comment by nikic** [comment]
Cherry-picked https://github.com/rust-lang/llvm-project/commit/e6c5dd473b7da567dc20ed3a635ca4dd2c6b7fb6 for the mingw profiler runtime fix and adjusted IR expectation for https://reviews.llvm.org/D103355. Let's check whether this makes the mingw build pass...

@bors try

**Comment by nikic** [comment]
Looks like that worked, let's try the full build again...

@bors r=nagisa

**Comment by klensy** [comment]
> 
> 
> The job **`x86_64-msvc-1`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/runs/3344612182) [(plain)](https://github.com/rust-lang-ci/rust/commit/de29cd3e8c931de2fa775393077b665cca0df1c1/checks/3344612182/logs)
> Click to see the possible cause of the failure (guessed by this bot)

This can be random error (at least not specified to this pr), saw something like that few times earlier.

**Comment by nikic** [comment]
Could retry that job, but I think it's pretty plausible that the LLVM upgrade broke coverage somehow. Unfortunately I don't have a Windows system to debug.

cc @richkadel

**Comment by nikic** [comment]
I tried to debug this in a Windows VM, but wasn't able to get run-make-fulldeps tests to work.

**Comment by nikic** [comment]
I got the tests to work in an msys2 environment. It looks like the switch to `llvm.compiler.used` is the problem here. Changing that back to `llvm.used` makes the coverage-reports test work again.

It looks like the clang implementation unconditionally uses `llvm.used` for coverage symbols, so I'll follow along with that.

**Comment by bstrie** [comment]
Is there an issue to track the bug that motivated the original switch to `llvm.compiler.used`?

**Comment by nikic** [comment]
> Is there an issue to track the bug that motivated the original switch to `llvm.compiler.used`?

You mean a tracking issue our side? I don't think there's anything useful to track there, because the issue is already fixed in gold, it's just old versions that are affected. Where "old" should be read as "not bleeding edge".

FWIW I also left a comment about this on the upstream patch (https://reviews.llvm.org/D97448#2933400), but I'm not sure the author understood the distinction between "old binutils versions ignore retain semantics" and "old binutils will silently miscompile code in ways that has no relation to retain semantics".

**Comment by nikic** [comment]
`@__llvm_profile_runtime_user` is getting placed in `@llvm.compiler.used` rather than `@llvm.used` on Windows as well, so drop the separate expectation for that...

@bors r=nagisa

**Comment by leonardo-m** [comment]
This has reduced by about 20% the number of array bound panics (and other similar panic) in my binaries (despite the final binary size is few percent bigger), this is a significant (good) change, I don't yet know what caused it but I'll try to understand it.

**Comment by adrian17** [comment]
> despite the final binary size is few percent bigger

I also observe 1-3% size increases in final binaries (including wasm); not sure if that is to be expected, or warrants an issue.

**Comment by leonardo-m** [comment]
> not sure if that is to be expected, or warrants an issue.

In my opinion this kind of slow little change is better tracked with specialized efforts (like the one for the compiler performance).

**Pull request merged**